[UserAccountsDrawerHeader] New option for changing profiles images size

burekas7 opened the pull request:
According to this issue: https://github.com/flutter/flutter/issues/73012
I made the changes and add  the option for changing profile image size.

burekas7 committed:
[UserAccountsDrawerHeader] New option for changing profiles images size

According to this issue: https://github.com/flutter/flutter/issues/73012
I made the changes and add  the option for changing profile image size.

burekas7 committed:
Merge pull request #1 from burekas7/burekas7-UserAccountsDrawerHeader-ImagesSize

[UserAccountsDrawerHeader] New option for changing profiles images size

burekas7 committed:
[UserAccountsDrawerHeader] New option for changing profiles images size

According to this issue: https://github.com/flutter/flutter/issues/73012
I made the changes and add  the option for changing profile image size.

burekas7 committed:
Merge pull request #2 from burekas7/burekas7-UserAccountsDrawerHeader-ImagesSize

[UserAccountsDrawerHeader] New option for changing profiles images size

burekas7 committed:
[UserAccountsDrawerHeader] New option for changing profiles images size

According to this issue: https://github.com/flutter/flutter/issues/73012
I made the changes and add the option for changing profile image size.

burekas7 committed:
Merge pull request #3 from burekas7/burekas7-UserAccountsDrawerHeader-ImagesSize

[UserAccountsDrawerHeader] New option for changing profiles images size

justinmc commented on packages/flutter/lib/src/material/user_accounts_drawer_header.dart:
```diff
@@ -20,10 +20,14 @@ class _AccountPictures extends StatelessWidget {
   final Widget? currentAccountPicture;
   final List<Widget>? otherAccountsPictures;
+  final double? currentAccountPictureSize;
+  final double? otherAccountsPicturesSize;
```
What do you think about making these `Size` classes instead of doubles?

justinmc replied:
Also, either way, they should probably default to the original values (48 and 72).

burekas7 replied:
> What do you think about making these `Size` classes instead of doubles?
What is the reason to make them classes, they are just numbers.
What do you mean by make them classes?

burekas7 replied:
> Also, either way, they should probably default to the original values (48 and 72).

I will check this.

burekas7 replied:
@justinmc 
By the way, what is the best practice to define default value, in the constructor definition or another way inside the widget where using the variables?

GaspardMerten replied:
Defining your default variables in the constructor is probably the most efficient way!

burekas7 replied:
@GaspardMerten
I have a basic question:
If Widget1 calls Widget2, where should I put the default values:
1. only at the Widget1 constructor?
2. only at the Widget2 constructor?
3. On both Widget1 and Widget2 constuctors?

GaspardMerten replied:
I would say that it depends whether or not Widget 2 is private or not. Meaning that if some user will one day directly use widget 2 then I think that yes there should be a default value in the constructor. If not, then since you are sure that Widget 2 will only be constructed by Widget 1, adding default values to Widget 2 would not bring any value. (My opinion :) )

burekas7 replied:
> What do you think about making these `Size` classes instead of doubles?

Done.

burekas7 replied:
> Also, either way, they should probably default to the original values (48 and 72).

Done.

burekas7 replied:
> I would say that it depends whether or not Widget 2 is private or not. Meaning that if some user will one day directly use widget 2 then I think that yes there should be a default value in the constructor. If not, then since you are sure that Widget 2 will only be constructed by Widget 1, adding default values to Widget 2 would not bring any value. (My opinion :) )

Indeed, makes sense.
Since the second widget is private, as you said, I put the defaults only in the first/main widget constructor which is public.

justinmc commented on packages/flutter/lib/src/material/user_accounts_drawer_header.dart:
```diff
@@ -321,6 +327,15 @@ class UserAccountsDrawerHeader extends StatefulWidget {
   final List<Widget>? otherAccountsPictures;
 
+  /// The size of the widget that placed in the upper-left corner that represents the current
+  /// user's account. Normally a [CircleAvatar].
```
The first line should be one simple sentence, so this should be more like this:

```
/// The size of the widget that placed in the upper-left corner that represents the current
/// user's account.
///
/// Normally a [CircleAvatar].
```

Same for the otherAccountsPicturesSize docs below.

burekas7 replied:
OK,
I will edit that according to your suggestion.

justinmc reviewed (commented):
Here's what you should do for a test.  Add a new test to [user_accounts_drawer_header_test.dart](https://github.com/flutter/flutter/blob/master/packages/flutter/test/material/user_accounts_drawer_header_test.dart) by copying one of the existing tests.  Render a UserAccountsDrawerHeader and find the two widgets whose size you are making configurable.  You might be able to find these by using `find.byType` for Container and SizedBox if we're lucky.  Expect that they have the default sizes.  Then pass in new sizes to them and expect that they have the given sizes.  You could do this in two different tests, or use a StatefulBuilder.

Let me know if you need any more specific help with any part of those tests.

justinmc commented:
Also, you have a test failure due to changing the default size I think:

```
The following TestFailure object was thrown running a test:
  Expected: <72.0>
  Actual: <40.0>

When the exception was thrown, this was the stack:
#4      main.<anonymous closure> (file:///b/s/w/ir/k/flutter/packages/flutter/test/material/user_accounts_drawer_header_test.dart:82:5)
<asynchronous suspension>
<asynchronous suspension>
(elided one frame from package:stack_trace)
...
```

burekas7 committed:
Update user_accounts_drawer_header.dart

Fix defaults sizes & Fix  variables descriptions

burekas7 commented:
@googlebot I signed it!

burekas7 commented:
> Here's what you should do for a test. Add a new test to [user_accounts_drawer_header_test.dart](https://github.com/flutter/flutter/blob/master/packages/flutter/test/material/user_accounts_drawer_header_test.dart) by copying one of the existing tests. Render a UserAccountsDrawerHeader and find the two widgets whose size you are making configurable. You might be able to find these by using `find.byType` for Container and SizedBox if we're lucky. Expect that they have the default sizes. Then pass in new sizes to them and expect that they have the given sizes. You could do this in two different tests, or use a StatefulBuilder.
> 
> Let me know if you need any more specific help with any part of those tests.

I'm very new in Flutter, so I still don't familiar with all the Test section.

burekas7 committed:
Merge pull request #4 from burekas7/burekas7-UserAccountsDrawerHeader-ImagesSize

Update user_accounts_drawer_header.dart

GaspardMerten commented:
If you fix everything that has already been said (default values, using Size and not double, ...). Then I can write the tests for you!

burekas7 committed:
Update user_accounts_drawer_header.dart

Change size variables type from double to Size class.

burekas7 committed:
Merge branch 'burekas7-UserAccountsDrawerHeader-ImagesSize'

burekas7 commented:
> If you fix everything that has already been said (default values, using Size and not double, ...). Then I can write tests for you!

I fixed what has been said (Default sizes, using the Size class and descriptions).
It would be nice that you helo me with the Tests.
Thank you.

burekas7 commented:
I have something very strange with the default value test.
When I'm set the default at the constructor and use it with "widget.variable_name"
The size of the SizedBox is null, but when I type manually the Size inside the build tree, SizedBox has a value.
So what is the difference, why using widet. isn't passing the value?

GaspardMerten commented:
Could you give us an example (code) of what you are talking about?

Commit:
Fixing incoherent size behavior.

When specifying a size of 48 for other accounts pictures, the size was reduced by 8 since the padding was applied directly to the container of which we defined the size with the other accounts pictures size parameter.

 Changed that behavior by adding an extra Padding widget on top of the Container which, since it only used width and height parameters, has been changed to a SizedBox.size

Commit:
Adding new test verifying that the default sizes can actually be changed + updating the sufficient tap targets size test

GaspardMerten commented:
Hi @burekas7, 

I wrote the tests for your PR  + some minor changes!

Could you give me access to your Flutter repo?

burekas7 commented:
@GaspardMerten 

According to previouse comment I focus on one of the properties:

For example:

Constructor:
```
  const UserAccountsDrawerHeader({
    this.currentAccountPictureSize = const Size.square(72.0),
  }) : super(key: key);
```

`final Size? currentAccountPictureSize;`

Using it inside the widget build:
`currentAccountPictureSize: widget.currentAccountPictureSize,`

When I run the tests, the SizedBox return null as the width value.
But when I replace it with fixed value like this:
`currentAccountPictureSize: const Size.square(72.0)`
It does return a value.

So why the default value in the constructor isn't set in the widget build?

burekas7 commented:
> Hi @burekas7,
> 
> I wrote the tests for your PR + some minor changes!
> 
> Could you give me access to your Flutter repo?

I added you.

GaspardMerten commented:
@burekas7 that's because you are giving a null value to the constructor inside pumpTestWidget.  So pumpTestWidget should look like that:

```
Future<void> pumpTestWidget(
    WidgetTester tester, {
      bool withName = true,
      bool withEmail = true,
      bool withOnDetailsPressedHandler = true,
      Size otherAccountsPictureSize = const Size.square(40),
      Size currentAccountPictureSize  = const Size.square(72),
    }) async {
```

I just pushed my commits on your repo! :)

GaspardMerten commented:
+ If you think my commits are useful, please add them to this PR :)

burekas7 commented:
> I just pushed my commits on your repo! :)

Thanks I will check this.

> @burekas7 that's because you are giving a null value to the constructor inside pumpTestWidget. So pumpTestWidget should look like that:
> 
> ```
> Future<void> pumpTestWidget(
>     WidgetTester tester, {
>       bool withName = true,
>       bool withEmail = true,
>       bool withOnDetailsPressedHandler = true,
>       Size otherAccountsPictureSize = const Size.square(40),
>       Size currentAccountPictureSize  = const Size.square(72),
>     }) async {
> ```
> 

@GaspardMerten 
But I didn't used the "pumpTestWidget" function, I wrote it like this (As it does in one of the tests):
Why does I need to set defaults in the Test class, the defaults are defined in the Widget class.
```
  testWidgets('UserAccountsDrawerHeader account profile size changes', (WidgetTester tester) async {
    Widget buildFrame({
      Widget? currentAccountPicture,
      List<Widget>? otherAccountsPictures,
      Size? currentAccountPictureSize,
      Size? otherAccountsPicturesSize,
      Widget? accountName,
      Widget? accountEmail,
      VoidCallback? onDetailsPressed,
      EdgeInsets? margin,
    }) {
      return MaterialApp(
        home: Material(
          child: Center(
            child: UserAccountsDrawerHeader(
              currentAccountPicture: currentAccountPicture,
              otherAccountsPictures: otherAccountsPictures,
              currentAccountPictureSize: currentAccountPictureSize,
              otherAccountsPicturesSize: otherAccountsPicturesSize,
              accountName: accountName,
              accountEmail: accountEmail,
              onDetailsPressed: onDetailsPressed,
              margin: margin,
            ),
          ),
        ),
      );
    }

    await tester.pumpWidget(buildFrame());

    SizedBox sizedBoxWidget = tester.firstWidget(find.byType(SizedBox));

    expect(sizedBoxWidget.width, 72.0);
    expect(sizedBoxWidget.height, 72.0);

  });
```

GaspardMerten commented:
Well, what I said is still applicable to your case, currentAccountPictureSize and othersAccountPicturesSize should never be null. So you could/should actually replace Size? by Size then specify the default value of those parameters as it is done inside the widget constructor that you wrote.

But the method you used should is redundant since there already is a pumpTestWidget function which simplifies the test's writing.

burekas7 commented:
> Well, what I said is still applicable to your case, currentAccountPictureSize and othersAccountPicturesSize should never be null. So you could/should actually replace Size? by Size then specify the default value of those parameters as it is done inside the widget constructor that you wrote.
> 
> But the method you used should is redundant since there already is a pumpTestWidget function which simplifies the test's writing.

So basically I should set defaults also in the Test class?
I was thinking the concept is that I need to set the defaults only in the Widget class,
so in the Test class I'm calling the widget and testing if the defaults were set after the widget has been created.

burekas7 commented:
> * If you think my commits are useful, please add them to this PR :)

Thanks.
How can I merge your commits with my files?

GaspardMerten commented:
> How can I merge your commits with my files?

I just did it :)

GaspardMerten commented:
> > Well, what I said is still applicable to your case, currentAccountPictureSize and othersAccountPicturesSize should never be null. So you could/should actually replace Size? by Size then specify the default value of those parameters as it is done inside the widget constructor that you wrote.
> > But the method you used should is redundant since there already is a pumpTestWidget function which simplifies the test's writing.
> 
> So basically I should set defaults also in the Test class?
> I was thinking the concept is that I need to set the defaults only in the Widget class,
> so in the Test class I'm calling the widget and testing if the defaults were set after the widget has been created.

Only because you are specifying a null argument, usually when you don't want to change the parameter you just do not specify it. 

What I am trying to say is that:

Widget(size: null) won't work because Widget() should be used.

GaspardMerten commented:
@googlebot I fixed it.

GaspardMerten commented:
@googlebot I fixed it.

GaspardMerten commented:
@googlebot I consent.

GaspardMerten commented:
@justinmc Does this PR looks good to you?

burekas7 commented:
@justinmc 
When we can use this pr with the main release?
Should I need to update something in my packages?

justinmc commented on packages/flutter/lib/src/material/user_accounts_drawer_header.dart:
```diff
@@ -36,14 +40,15 @@ class _AccountPictures extends StatelessWidget {
-                child: Semantics(
-                  container: true,
-                  child: Container(
+                child: Padding(
```
Could these two Padding widgets be combined?

GaspardMerten replied:
We could write `const EdgeInsets.only(left: 8.0, bottom: 8.0).add(const EdgeInsetsDirectional.only(start: 8.0))`

But I feel like this would be a bit too unreadable.

Are we sure that we can not merge both of them into one single EdgeInsetsDirectional? I am not aware of the design guidelines for this widget. But if yes we could end up with just `const EdgeInsetsDirectional.only(start: 16.0, end: 8.0)`

justinmc replied:
Oh I see, I missed that it was using both `left` and `start`.  I think it's alright to keep this as-is with two Padding widgets for now, since that's effectively how it was before this PR.

justinmc commented on packages/flutter/lib/src/material/user_accounts_drawer_header.dart:
```diff
@@ -299,6 +303,8 @@ class UserAccountsDrawerHeader extends StatefulWidget {
     this.currentAccountPicture,
     this.otherAccountsPictures,
+    this.currentAccountPictureSize = const Size.square(72.0),
+    this.otherAccountsPicturesSize = const Size.square(40.0),
```
Was the default for otherAccountsPicturesSize 48px originally?  Looking at the deleted code above.  Or is it now affected differently by the padding?

GaspardMerten replied:
A padding of 8px has been moved above the SizedBox related to the otherAccountsPicturesSize.  So the actual size is the same.

justinmc replied:
Ah got it.

justinmc commented on packages/flutter/test/material/user_accounts_drawer_header_test.dart:
```diff
@@ -13,11 +13,13 @@ const Key avatarC = Key('C');
+      bool withName = true,
+      bool withEmail = true,
+      bool withOnDetailsPressedHandler = true,
+      Size otherAccountsPictureSize = const Size.square(40),
+      Size currentAccountPictureSize  = const Size.square(72),
```
These two should have the `.0`, like `40.0` in order to be consistent with above.

justinmc replied:
Or maybe they don't need defaults since UserAccountsDrawerHeader will set these defaults anyway?

GaspardMerten replied:
I do think we need them, that is if we want to be able to use the pumpTestWidget function for testing different size. If we don't specify them we get an error because of null-safety. I don't really know what is the best practice for this case...

justinmc replied:
Sounds good, keeping it like this is ok but add the `.0`.

GaspardMerten replied:
Done :) ! @justinmc

justinmc reviewed (commented):
Sorry for the slow response here!  Just some small comments.

Commit:
Replacing float in int format by only float format for better consistency

justinmc commented:
Looks like you've got some analyzer failures, can you look at those?

GaspardMerten commented:
@justinmc I don't really see what could have affected the tests' results with my last commit. I literally just replaced x with x.0... 

Could it be possible that the issue comes from "above"? Or am I missing something?

GaspardMerten commented:
@justinmc, I checked the other current PR tests results and they all share the same results like the one we have. So since my last commit wasn't very sensitive.. (I replaced x with x.0). So I think that it is the master tree which is not currently sane.  Or am I missing something?

burekas7 committed:
Merge pull request #5 from burekas7/burekas7-UserAccountsDrawerHeader-ImagesSize

Replacing float in int format by only float format for better consist…

justinmc commented:
Ah you're right.  Could you try merging in master?  Hopefully that will fix it.

Commit:
Merge branch 'master' of https://github.com/flutter/flutter into burekas7-UserAccountsDrawerHeader-ImagesSize

GaspardMerten commented:
It seems like the tree is still not sane :(, I will try again tomorrow.

Commit:
Merge branch 'master' of https://github.com/flutter/flutter into burekas7-UserAccountsDrawerHeader-ImagesSize

Commit:
Merge branch 'master' of https://github.com/flutter/flutter into burekas7-UserAccountsDrawerHeader-ImagesSize

GaspardMerten commented:
@justinmc tree is green :) !

justinmc reviewed (approved):
LGTM 👍 

I'll try to find a second reviewer for this.

shihaohong commented on packages/flutter/test/material/user_accounts_drawer_header_test.dart:
```diff
@@ -528,17 +549,17 @@ void main() {
     expect(tester.getSemantics(find.text('B')), matchesSemantics(
       label: 'B',
-      size: const Size(48.0, 48.0),
+      size: const Size(40.0, 40.0),
```
Shouldn't the size of the tap targets still be 48.0 by 48.0, here and below? I think that's the minimum required size for mobile right?

shihaohong replied:
https://github.com/flutter/flutter/blob/master/packages/flutter/lib/src/material/theme_data.dart#L93

GaspardMerten replied:
I changed the Semantic widget place, which made the semantic widget returns to its original size of 48.0.  But the size of the children  (otherAccountsPicturesSize )  still is 40.0 because of the surrounding padding. That is what was originally done (before this PR), are we sure that's the right behaviour/design/guideline :) ?

shihaohong commented on packages/flutter/lib/src/material/user_accounts_drawer_header.dart:
```diff
@@ -35,14 +39,15 @@ class _AccountPictures extends StatelessWidget {
-                    height: 48.0,
-                    child: picture,
-                 ),
+                  child: Semantics(
```
Shouldn't the semantics wrap one of the paddings as well to reach the minimum tap target size of 48x48 on mobile?

GaspardMerten replied:
Yep, fixed!

burekas7 commented:
@GaspardMerten @justinmc 
Thank you!

Commit:
Updating semantic widget size to minimum requirements (48.0 instead of 40.0 dp)

shihaohong commented on packages/flutter/test/material/user_accounts_drawer_header_test.dart:
```diff
@@ -104,6 +108,23 @@ void main() {
+    const Size currentAccountPictureSize = Size.square(60.0);
+    const Size otherAccountsPictureSize = Size.square(30.0);
+
+    await pumpTestWidget(
+        tester,
+        currentAccountPictureSize: currentAccountPictureSize,
+        otherAccountsPictureSize: otherAccountsPictureSize
+    );
```
nit:
```suggestion
    await pumpTestWidget(
      tester,
      currentAccountPictureSize: currentAccountPictureSize,
      otherAccountsPictureSize: otherAccountsPictureSize,
    );
```

shihaohong reviewed (approved):
LGTM with nit, thanks for fixing the semantics issue!

AlexV525 commented on packages/flutter/test/material/user_accounts_drawer_header_test.dart:
```diff
@@ -13,11 +13,13 @@ const Key avatarC = Key('C');
-  bool withEmail = true,
-  bool withOnDetailsPressedHandler = true,
-}) async {
+    WidgetTester tester, {
```
Seems not aligned at here? (All arguments)

```suggestion
  WidgetTester tester, {
```

GaspardMerten replied:
IMO it is quite normal since 'tester' is not a named argument but a positioned one.

AlexV525 commented on packages/flutter/lib/src/material/user_accounts_drawer_header.dart:
```diff
@@ -181,8 +185,8 @@ class _AccountDetailsState extends State<_AccountDetails> with SingleTickerProvi
                       semanticLabel: widget.isOpen
-                        ? localizations.hideAccountsLabel
-                        : localizations.showAccountsLabel,
+                          ? localizations.hideAccountsLabel
```
IMO we should revert all unnecessary format changes

AlexV525 replied:
Still needs a revert at here.

AlexV525 commented on packages/flutter/lib/src/material/user_accounts_drawer_header.dart:
```diff
@@ -37,12 +41,13 @@ class _AccountPictures extends StatelessWidget {
-                    child: picture,
-                 ),
+                  child: Padding(
+                    padding: const EdgeInsets.only(left: 8.0, bottom: 8.0),
```
Is this padding following any material design spec? If not, can it be customizable? (I'm pretty new to to this widget)

GaspardMerten replied:
Well that should be for an other PR...

GaspardMerten committed:
Update packages/flutter/test/material/user_accounts_drawer_header_test.dart

GaspardMerten committed:
Update packages/flutter/test/material/user_accounts_drawer_header_test.dart

GaspardMerten commented:
Are we good to merge? :)

AlexV525 commented on packages/flutter/test/material/user_accounts_drawer_header_test.dart:
```diff
@@ -13,11 +13,13 @@ const Key avatarC = Key('C');
-  bool withEmail = true,
-  bool withOnDetailsPressedHandler = true,
-}) async {
+    WidgetTester tester, {
+      bool withName = true,
+      bool withEmail = true,
+      bool withOnDetailsPressedHandler = true,
+      Size otherAccountsPictureSize = const Size.square(40.0),
+      Size currentAccountPictureSize  = const Size.square(72.0),
+    }) async {
```
```suggestion
  WidgetTester tester, {
    bool withName = true,
    bool withEmail = true,
    bool withOnDetailsPressedHandler = true,
    Size otherAccountsPictureSize = const Size.square(40.0),
    Size currentAccountPictureSize  = const Size.square(72.0),
  }) async {
```

GaspardMerten commented:
@AlexV525  could you accept the google-cla  :) ?

AlexV525 commented:
> @AlexV525 could you accept the google-cla :) ?

You can remove co-authored from commit message to avoid the CLA issue when commit suggestions. I've accepted it already.

GaspardMerten commented:
> > @AlexV525 could you accept the google-cla :) ?
> 
> You can remove co-authored from commit message to avoid the CLA issue when commit suggestions. I've accepted it already.

Oh I did not know that :) Thank you!

justinmc commented:
Thanks everyone!

burekas7 commented:
Thank you all.
Well done!

GaspardMerten commented:
Nice 😃
